jenkins-pillage.py

```diff
@@ -1,107 +1,153 @@
     parser = argparse.ArgumentParser(
-        description='Pillage sensitive information from Jenkins servers')
+        description="Pillage sensitive information from Jenkins servers"
+    )
     parser.add_argument(
-        '-b','--buildurl',
-        dest='build_url',
-        help='The build URL to pillage'
-        )
+        "-b", "--buildurl", dest="build_url", help="The build URL to pillage"
+    )
     parser.add_argument(
-        '-u','--user',
-        dest='username',
-        help='The Basic Auth username for the service'
-        )
+        "-u",
+        "--user",
+        dest="username",
+        help="The Basic Auth username for the service",
+        required=False,
+        default=None,
+    )
     parser.add_argument(
-        '-p','--password',
-        dest='password',
-        help='The Basic Auth password for the service'
-        )
+        "-p",
+        "--password",
+        dest="password",
+        required=False,
+        default=None,
+        help="The Basic Auth password for the service",
+    )
     parser.add_argument(
-        '-l','--list',
-        dest='list_url',
-        help='Lists all found build URLs to use with -b'
-        )
+        "-l",
+        "--list",
+        dest="list_url",
+        help="Lists all found build URLs to use with -b",
+    )
     parser.add_argument(
-        '-a','--auto',
-        dest='auto',
-        help='Automatically perform -l and then -b on a root URL'
-        )
+        "-a",
+        "--auto",
+        dest="auto",
+        help="Automatically perform -l and then -b on a root URL",
+    )
     parser.add_argument(
-        '-f','--force',
-        dest='force',
-        action='store_true',
-        help='Force the URLs to use the supplied socket (in case the server returns localhost)'
+        "-f",
+        "--force",
+        dest="force",
+        action="store_true",
+        help="Force the URLs to use the supplied socket (in case the server returns localhost)",
+    )
+    args = parser.parse_args()
+    if args.username:
+        auth = HTTPBasicAuth(
+            args.username, args.password or input("Please provide your password: ")
         )
-    args = parser.parse_args() 
-    if args.username and args.password:
-        auth = HTTPBasicAuth(args.username, args.password)
     else:
         auth = None
     if args.auto:
         print("Getting a list of all build URLs")
         build_links = get_all_build_links(args.auto, auth=auth, netloc_force=args.force)
-        for url in build_links:
-            build = JenkinsJobBuild(url, auth=auth)
-            print("Attempting: {}".format(build.url))
-            write_out_console_output(build)
-            write_out_zip_urls(build)
-            write_out_env_vars(build)
+        with Pool(processes=int(cpu_count() - 1) or 1) as pool:
+            pool.starmap(goodies, [(link, auth) for link in build_links])
+        decrypt_all_secrets(JenkinsConnection(args.auto, auth=auth))
```

Comment:
During some testing I found that running "-a https://example.com/" works properly, but fails if the user supplies "-a https://example.com". It looks like it's because the JenkinsJobBuild/JenkinsConnection validates there is a path appended to the URL in the __init__. I originally placed that check because the supplied URLs to JenkinsJobBuild ideally should only have been build URLs themselves, which always have a path component to them.

I'm kinda thinking it would be best to have a separate Jenkins class (JenkinsConnection) to handle all connection details/methods that could be composited into a JenkinsJobBuild class to better model how it all works. 

Alternatively you could also just implement a check for a path component in args.auto before passing it to JenkinsConnection and that would solve that problem as well lol.

Thoughts?

Reply:
Fixed with https://github.com/DolosGroup/Jenkins-Pillage/pull/1/commits/074e1149bf21a1dba75e60cb33b572b7ed2fe7ba

I like the composition idea in the future but to control scope went with appending the slash at the end of the URL.  Is that okay?  I tested locally with and without the slash when invoking the script and both work now.